cellrank/tl/kernels/_velocity_schemes.py

```diff
@@ -0,0 +1,345 @@
+# -*- coding: utf-8 -*-
+from abc import ABC, ABCMeta, abstractmethod
+from typing import Tuple
+from functools import partial
+
+import numpy as np
+from numba import njit
+
+from cellrank.ul._docs import d
+from cellrank.ul._utils import valuedispatch
+from cellrank.tl._constants import ModeEnum
+from cellrank.tl.kernels._utils import norm, np_mean, jit_kwargs
+
+
+class Scheme(ModeEnum):  # noqa: D101
+    DOT_PRODUCT = "dot_product"
+    COSINE = "cosine"
+    CORRELATION = "correlation"
+
+
+try:
+    import jax.numpy as jnp
+    from jax import jit, hessian
+
+    @jit
+    def _softmax_jax(x: np.ndarray, softmax_scale) -> np.ndarray:
+        numerator = x * softmax_scale
+        numerator = jnp.exp(numerator - jnp.max(numerator))
+        return numerator / jnp.sum(numerator)
+
+    @jit
+    def _softmax_masked_jax(
+        x: np.ndarray, mask: np.ndarray, softmax_scale
+    ) -> np.ndarray:
+        numerator = x * softmax_scale
+        numerator = jnp.exp(numerator - jnp.nanmax(numerator))
+        numerator = jnp.where(mask, 0, numerator)  # essential
+
+        return numerator / jnp.nansum(numerator)
+
+    @partial(jit, static_argnums=(3, 4))
+    def _predict_transition_probabilities_jax(
+        X: np.ndarray,
+        W: np.ndarray,
+        softmax_scale: float = 1.0,
+        center_mean: bool = True,
+        scale_by_norm: bool = True,
+    ):
+        if center_mean:
+            # pearson correlation, otherwise cosine
+            W -= W.mean(axis=1)[:, None]
+            X -= X.mean()
+
+        if scale_by_norm:
+            denom = jnp.linalg.norm(X) * jnp.linalg.norm(W, axis=1)
+            mask = jnp.isclose(denom, 0)
+            denom = jnp.where(jnp.isclose(denom, 0), 1, denom)  # essential
```

Comment:
So the denominator is zero if either we have zero velocity or zero displacement. In that case, we set the denominator to 1. What's the intuition behind that? If I have a zero velocity vector for a given cell, wouldn't that just mean it stays where it is? Not sure.

Reply:
@michalk8 ?

Reply:
@Marius1311 the masking is done to get the gradients flowing (they can't flow through NaN).
In case of 0 velocity vector, we will then get uniform probabilities across the neighbors (not sure if cells `x` is considering itself as a neighbour in the velo graph, but if yes, we could change it so that the self transition is 1).
In any case, the values in the numerator are ignored when computing the softmax: https://github.com/theislab/cellrank/pull/447/files/5b3a3bf113b77255de5e2676089f68414766de08#diff-955d7bd9b19c6ac54b6d0d65fd4a5fcd496b34d470f28f002637f7bad061e936R91
(i.e. in the case of displacement, if there are cells that are the same [displ. 0, but possible different velocities], the transition probability to those will be 0).

Reply:
Hi Mike, this makes sense. Cell `x` is not considered to be a neighbor of itself. The current behaviour (uniform over neighbors if velocity vector is zero) makes sense to me. It won't matter in practice as this should be a 0-probability event and only happen in exceptional cases. The other case, zero displacement, also makes sense to me. Fine!